ls1intum/Artemis
# [Feature] Time Locking of Complaints / More Feedback Requests

### Checklist

### Motivation and Context


This closes #2514

Unresolved complaints are now time locked for a certain time to the teaching assistant who first opened the complaint. **Before merging to production this will be set to 24h, currently it is set to 5 minutes to allow testing**. During the lock duration no other teaching assistants can resolve a complaint. **Only the first reviewer AND instructors can open a locked complaint**. After the lock time is over, another teaching assistant can open the complaint and thus aquire a new lock and so on. **The original reviewer can also manually give up his lock by clicking on the corresponding button**


### Description


The lock mechanism is implemented internally in the following way:

- A "lock" is internally represented by an unsubmitted complaint response. The lock duration is calculated from the createdDate of this entity + duration of the lock (for example 5 minutes). The owner of the lock is the "reviewer" of this unsubmitted complaint response.
- This unsubmitted complaint response is deleted and exchanged if the lock is refreshed.
- When the complaint is resolved, this unsubmitted complaint gets filled with the data using a PUT request.


### Steps for Testing


This is a major change of the whole complaint | more feedback logic. 

It should be tested that the following scenarios still work:

@Testers: In theory it should be enough the test one column here (for example only TextExercises) as a lot of components are reused. So dont stress yourself too much if you do not have enough time. 

|                      | TextExercise   | ProgrammingExercise | FileUploadExercise | ModellingExercise |
|----------------------|----------------|---------------------|--------------------|-------------------|
| Complaints ACCEPT    | COURSE \| EXAM | COURSE \| EXAM      | COURSE \| EXAM     | COURSE \| EXAM    |
| Complaints REJECT    | COURSE \| EXAM | COURSE \| EXAM      | COURSE \| EXAM     | COURSE \| EXAM    |
| MoreFeedbackRequests | COURSE \| EXAM | COURSE \| EXAM      | COURSE \| EXAM     | COURSE \| EXAM    |
|                      |                |                     |                    |                   |

In addition the locking feature should be tested
- Locking should work
- Locks should run out after 5 minutes
- Locks can be given up manually
- Instructors should not be affected by locks



### Test Coverage



Test Coverage of Files with Major changes:

ComplaintResponseService.java 93%
ComplaintResponseResource.java 95%

complaint-response.service.ts 93%
complaints-for-tutors.componen.ts 84%

### Screenshots


**LOGGED IN AS ARTEMIS_TEST_USER_6**
![image](https://user-images.githubusercontent.com/29718932/103748031-ce3a8e00-5003-11eb-8e60-593f4bfa6966.png)
![image](https://user-images.githubusercontent.com/29718932/103748124-ed392000-5003-11eb-9e1b-ce220d574b19.png)
![image](https://user-images.githubusercontent.com/29718932/103748175-ffb35980-5003-11eb-8812-e8425efc1a5d.png)
![image](https://user-images.githubusercontent.com/29718932/103748286-2ec9cb00-5004-11eb-9d15-ddf1944bb78f.png)

**LOGGED IN AS ARTEMIS_TEST_USER_7**
![image](https://user-images.githubusercontent.com/29718932/103748338-443ef500-5004-11eb-8a21-c1424c594782.png)
![image](https://user-images.githubusercontent.com/29718932/103748419-60429680-5004-11eb-8fce-c9f39fea9e26.png)

**NOW LETS WAIT UNTIL THE LOCK HAS RUN OUT**
![image](https://user-images.githubusercontent.com/29718932/103748629-b0215d80-5004-11eb-923c-fed63e09483a.png)
![image](https://user-images.githubusercontent.com/29718932/103748735-d8a95780-5004-11eb-9f17-c746e670845f.png)
![image](https://user-images.githubusercontent.com/29718932/103748756-e1019280-5004-11eb-8aeb-b1b78073572d.png)
![image](https://user-images.githubusercontent.com/29718932/103748796-ed85eb00-5004-11eb-92b3-95ec6c08a469.png)
